src/main/java/week3/WordSearch.java

```diff
@@ -0,0 +1,195 @@
+
+                    if (dictionary.isWord(newPrefix)) {             //O(|words|)
+                      words.add(newPrefix);                         //O(1)
+                    }
+
+                    return dictionary.isPrefix(newPrefix);          //O(|words|)
+                  })
+                  .collect(Collectors.groupingBy(c -> p + grid[c.x][c.y]))
+          )
+      );
+
+      prefixes = newPrefixes;
+    }
+
+    return words;
+  }
+
+  /*
+   * Method returns all valid adjacent coordinates of c w.r.t grid and current path
+   *
+   * All Paths returned:
+   *  1) Append a new cell to the end of path c
+   *  2) New cell has not been visited
+   *  3) New cell is within the grid
+   *  3) New cell is adjacent to the last cell in path c
+   *
+   *  N.B not test independently because of Path
+   *
+   *  Time Complexity : O(1) for loops do a constant number of iterations and
+   *  all operations either run in constant time (if, bitset check) or are negligible
+   *  (coordinates.add as arraylist will at most have size 8)
+   */
+  List<Path> getUnvisitedAdjCells(Path c) {
+
+    if (c == null) {
+      return null;
+    }
+
+    List<Path> coordinates = new ArrayList<>();
+
+    for (int x = -1; x <= 1; x++) {
+      for (int y = -1; y <= 1; y++) {
+
+        int newX = c.x + x;
+        int newY = c.y + y;
+
+        if (newX >= 0 && newX < rows
+            && newY >= 0 && newY < cols
+            && !c.locs.get(newY * rows + newX)) {
+
+          coordinates.add(new Path(newX, newY, (BitSet) c.locs.clone()));
+        }
+
+      }
+    }
+
+    return coordinates;
+  }
+
+  /*
+   * Initialize the Map with all paths/prefixes of length 1.
+   * Each entry
+   * 1) the key is found in the grid and isPrefix to some word in the dictionary
+   * 2) the entry all locations of the key
+   *
+   * Time Complexity: O(|grid|*|words|)
+    *   where |grid| is the size of your grid (rows*cols) / time complexity of the nested for loop
+    *         |words| is the size of your dictionary / time complexity of isPrefix
+   */
+  Map<String, List<Path>> initializeMap(Dictionary dict, char[][] grid) {
+
+    if (dict == null || grid == null) {
+      return null;
+    }
+
+    rows = grid.length;
+    cols = grid.length > 0 ? grid[0].length : 0;
+
+    Map<String, List<Path>> letters = new HashMap<>();
+
+    //empty string is a prefix of all words by default
+    for (int r = 0; r < rows; r++) {
+      for (int c = 0; c < cols; c++) {
+        String prefix = String.valueOf(grid[r][c]);
+
+        if (dict.isPrefix(prefix)) {
+          letters.putIfAbsent(prefix, new ArrayList<>());
+          letters.get(prefix).add(new Path(r, c));
+        }
+      }
+    }
+
+    return letters;
+  }
+
+  private class Path {
+
+    public int x;
+    public int y;
+    public BitSet locs = new BitSet(rows * cols);
+
+    public Path(int x, int y) {
+      this.x = x;
+      this.y = y;
+      this.locs.set(y * rows + x);
+    }
```

Comment:
To be clear, I meant something more like:

Constructor(a, b) {
  Constructor(a, b, new OtherThing());
}

Constructor(a, b, c) { ...

This is a pretty common pattern to avoid copying and pasting code between constructors that do similar things.